services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -489,6 +519,18 @@ private void checkMtlsHoKToken(TokenManager.AccessTokenResponseBuilder responseB
         }
     }
 
+    private void checkDPoPToken(TokenManager.AccessTokenResponseBuilder responseBuilder, boolean useRefreshToken) {
+        // KEYCLOAK-15169 OAuth 2.0 Demonstrating Proof-of-Possession at the Application Layer (DPoP)
+        // https://tools.ietf.org/id/draft-ietf-oauth-dpop-04.html#section-6
+        if (clientConfig.isDPoPEnabled()) {
+            DPoPUtil.bindToken(responseBuilder.getAccessToken(), dPoP);
+            // TODO do not bind refresh tokens issued to confidential clients, see 5. DPoP Access Token Request
```

Comment:
As this comments say and [DPoP's specification](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop#section-5) described, it might be better not to make refresh tokens to be DPoP bind for a confidential client.

Reply:
This one turned out to be particularly tricky. In `TokenManager#refreshAccessToken()`, we're copying the `cnf` section from the refresh token to the newly issued access token. This won't work if the refresh token doesn't contain `cnf` anymore, e.g. in case of confidential client. Thus, we'll need to recompute the bindings anew. To keep the code aligned, I've moved recomputation to `TokenEndpoint#refreshTokenGrant()`, similar to what we're doing for authorization code and ROPC. I've also removed `TokenManager.RefreshResult`, which seems to be an ad-hoc class used only once. Instead, I've made `TokenManager#refreshAccessToken()` return a mutable `AccessTokenResponseBuilder`.

Please let me know if you think it would be better to remove `TokenManager.RefreshResult` via a separate PR.

Reply:
@dteleguin IMO, it might be better to treat this matter as a separate PR.